standard/types.md

```diff
@@ -804,14 +806,14 @@ When the nullable context is annotations:
 
 #### §Nullable-Warning-Context Nullable warnings
 
-When the nullable context is warnings:
+When the nullable context is ***warnings***:
 
 The compiler uses static flow analysis to determine the null state of any reference variable. When nullable warnings are enabled, a reference variable’s null state (§Nullabilities-And-Null-States) is either *not null*, *maybe null*, or *maybe default* and
 
 - A warning shall be generated if a reference variable that has been determined to be *maybe null*, is dereferenced.
-- The state of a reference variable is *maybe null* unless the compiler can determine one of two conditions:
+- The state of a reference expression is *maybe null* unless the compiler can determine one of two conditions:
   1. The variable has been definitely assigned a non-`null` value.
-  1. The variable or expression has been checked against `null` before dereferencing it.
+  1. The variable has been checked against `null` before dereferencing it.
```

Comment:
Expression for this and the line above?

Reply:
I think this has to be "variable", not expression. It's for the same reason as https://github.com/dotnet/csharpstandard/pull/1124/files#r1626266110

However, changing to a property removes the warnings:

See [Sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMAGAWAChcBGY4gNwmQAIBjWgXlpgFMB3WgYQAoBKANzEoAM1q96AOgCytKAGdWAewStUAGw39itPbSykAnJNlSAMmxgBzBAAshtAPRPaAOWW0ONGLGtSKImo6JRZpGWEiMQklRRU1GE1tXX1DEwULK1sHQWdXDy8fPwCiYiwAZgMcHmIAbxS9CoNSPAB+Wjla2ms2BFyFXtyAX2IhoA==)

Reply:
In-meeting discussion: we can probably use variables for now, but revisit later. Some options:

- Define what a "dotted sequence of variables then field/property names" is and use that
- Talk about "expressions in general" but without expecting compilers to track *all* expressions (but go into no further detail)
- Talk about variables specifically, but give compilers enough options to add extra diagnostics (without being specific about it being related to null state tracking)

Happy to talk about that later when we know more about the shape of things.

Reply:
Listening to the meeting, I am coming to this conclusion for expressions vs. variables:

- The result of evaluating an expression has a *null state*. In other words, there's a known null state for any expression and known operands.
- Variables have a *null state* associated with them on any line of source code. 

(Need to wordsmith, but the idea is that a variable has a null state, and any number of operations like assignment, or performing a null check impacts the null state of the variable. That same concept won't apply to an expression. For example:

```csharp
// declaration:

object? M() => null;

var n = M();
if (n is not null)
{
    // null state of n is not null
} 
if (M()) is not null)
{
    var v = M();
    // null state of v is Maybe null.
}
```

}

Reply:
Extending the last example:

```csharp
if (n is null)
{
   // null state of n is Maybe null
}
```

The “maybe null” is at least curious… the issue is more than the null state of variables vs. expressions; it is the available null states themselves, or at least how they are described/explained.

We’re developing this collection of PRs to answer questions as well as coming up with possible wording (and consequentially we shouldn’t merge any of them until we’re happy to merge all of them).

A fundamental question which I don’t think we know the answer to yet is whether the approach of not specifying the algorithm that determines null states, while limiting it to determine only 2 out of the possible 3 states (is null, is not null, unknown), is that the right balance between specifying too little and too much?

Unless we do have our answer to that question?